initfiles/bash/etc/init.d/setupPKI.in

```diff
@@ -0,0 +1,128 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+################################################################################
+
+###<REPLACE>###
+
+progname=setupPKI
+
+source ${INSTALL_DIR}/etc/init.d/hpcc_common
+source ${INSTALL_DIR}/etc/init.d/init-functions
+source  ${INSTALL_DIR}/etc/init.d/export-path
+
+HPCC_CONFIG=${HPCC_CONFIG:-${CONFIG_DIR}/${ENV_CONF_FILE}}
+SECTION=${1:-DEFAULT}
+PATH_PREFIX=`cat ${HPCC_CONFIG} | sed -n "/\[${SECTION}\]/,/\[/p" | grep "^home *= *" | sed -e 's/^home *= *//'`
+USER_NAME=`cat ${HPCC_CONFIG} | sed -n "/\[${SECTION}\]/,/\[/p" | grep "^user *= *" | sed -e 's/^user *= *//'`
+homePath=${PATH_PREFIX}/${USER_NAME}
+
+certPath=${homePath}/certificate
+sbin_path="${INSTALL_DIR}/sbin"
+envfile="${CONFIG_DIR}/${ENV_XML_FILE}"
+
+COMPS=$(${sbin_path}/configgen -env ${envfile} -list)
+
+if [[ ${rc} -ne 0 ]]; then
+    log "${progname}: failure to build COMPS from configgen call"
+    echo -e "\033[31merror\033[0m: ${progname} -> failure to build COMPS from configgen call"
+    exit 1
+fi
+
+comp.parser ${COMPS}
+
+if [ -z ${compArray} ];then
+   log  "There are no components configured to run on this node ..."
+   echo "There are no components configured to run on this node ..."
+   exit 3
+fi
+
+if [ ! -d ${certPath} ]; then
+    mkdir -p ${certPath}
+fi
+
+domainname=${DOMAINNAME}
+if [ "$domainname" = "" ]; then
+    domainname=local
+fi
+
+subjbase="/C=US/ST=FL/L=Boca Raton/O=HPCCSystems/OU=Customer"
+
+regenerate=0
+if [ ! -e ${certPath}/cacert-key.pem ] || [ ! -e ${certPath}/cacert.pem ]; then
+    rm -rf ${certPath}/cacert-key.pem ${certPath}/cacert.pem
+    subj="${subjbase}/CN=ca.${domainname}"
+    openssl req -nodes -newkey rsa:2048 -keyout ${certPath}/cacert-key.pem -out ${certPath}/cacert.csr -subj "${subj}"
+    chmod 400 ${certPath}/cacert-key.pem
+    openssl x509 -req -days 365 -in ${certPath}/cacert.csr -signkey ${certPath}/cacert-key.pem -sha256 -out ${certPath}/cacert.pem
+    rm ${certPath}/cacert.csr
+    printf "Generated self-signed CA certificate and privatekey.\n"
+    regenerate=1
+fi
+
+for i in "" ${compArray[@]} ; do
+    compName=""
+    if [ "$i" != "" ]; then
+        compName=""${i#*_}
+    fi
+    if [ "${compName}" != "" ]; then
+         compNamePrint=${compName}
+    else
+         compNamePrint="\"\" (default)"
+    fi
+    if [ ! -d ${certPath}/${compName} ]; then
+        mkdir -p ${certPath}/${compName}
+    fi
+
+    if [ "${regenerate}" = "1" ] || [ ! -e ${certPath}/${compName}/key.pem ] || [ ! -e ${certPath}/${compName}/certificate.pem ]; then
+        if [ "${compName}" != "" ]; then
+             cn="${compName}.${domainname}"
+        else
+             cn="hpcc.${domainname}"
```

Comment:
I suppose if we get mutual authentication right, and a cn ever ends up starting with "hpcc." and something tries to use it, then identification should fail because it would never match what the other endpoint expected.

Reply:
Yes that's correct. This default certificate is mostly for backward compatibility purpose, so that the systems still runs before various components make the change to use their own certificates.